File: javascript/beer-song/beer-song.js
const oneTheWall = (bottles) => {
  if (bottles > 1) {
    return `${bottles} bottles of beer on the wall, ${bottles} bottles of beer.`;
  }
  if (bottles === 1) {
    return `1 bottle of beer on the wall, 1 bottle of beer.`;
  }
  if (bottles === 0) {
    return "No more bottles of beer on the wall, no more bottles of beer.";
  }
  throw new Error("Bad input");
};

const takeOneDown = (bottles) => {
  if (bottles > 2) {
    return `Take one down and pass it around, ${
      bottles - 1
    } bottles of beer on the wall.`;
  }
  if (bottles === 2) {
    return "Take one down and pass it around, 1 bottle of beer on the wall.";
  }
  if (bottles === 1) {
    return "Take it down and pass it around, no more bottles of beer on the wall.";
  }
  if (bottles === 0) {
    return "Go to the store and buy some more, 99 bottles of beer on the wall.";
  }
  throw new Error("Bad input");
};

export const recite = (initialBottlesCount, takeDownCount) => {
  let song = [];
  const endBottlesCount = initialBottlesCount - takeDownCount + 1;
  for (let i = initialBottlesCount; i >= endBottlesCount; i--) {
    song.push(oneTheWall(i), takeOneDown(i));
    if (i > endBottlesCount) {
      song.push("");
    }
  }
  return song;
};

File: elixir/beer-song/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(number) when number > 2,
    do: """
    #{number} bottles of beer on the wall, #{number} bottles of beer.
    Take one down and pass it around, #{number - 1} bottles of beer on the wall.
    """

  def verse(number) when number === 2,
    do: """
    2 bottles of beer on the wall, 2 bottles of beer.
    Take one down and pass it around, 1 bottle of beer on the wall.
    """

  def verse(number) when number === 1,
    do: """
    1 bottle of beer on the wall, 1 bottle of beer.
    Take it down and pass it around, no more bottles of beer on the wall.
    """

  def verse(number) when number === 0,
    do: """
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    """

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(), do: lyrics(99..0)

  def lyrics(range) do
    range |> Stream.map(&verse/1) |> Enum.join("\n")
  end
end
